docs/about/coding_guidelines.md

```diff
@@ -0,0 +1,68 @@
+---
+id: about_coding_guidelines
+title:  "ZIO Coding Guidelines"
+---
+
+These are coding guidelines strictly for ZIO contributors for ZIO projects and 
+not general conventions to be applied by the Scala community.
+
+Additionally, bear in mind that, although we try to enforce these rules to the 
+best of our ability, both via automated rules (scalafix) and strict reviewing 
+processes, it is both possible to find existing code that does not comply to 
+these rules. If that is the case, we would be extremely grateful if you could 
+make a contribution, by proving a fix to said issue.
+
+Last, but not least, these rules are continuously evolving and as such, 
+refer to them once in a while when in doubt.
+
+
+### Defining classes and traits
+
+1. Value classes must be final and extend AnyVal.
+2. Method extension classes must be final and extend `AnyVal`.
+3. Sealed traits that are ADTs should extend `Product` and `Serializable`.
+4. Regular traits and sealed trait that do not form ADTS should extend `Serializable` but not `Product`.
+5. Multiple parameter classes that are ADT should be case class and not extend AnyVal
+6. Traits should always extends `Serializable`. (i.e. `ZIO`)
+
+### Final and private modifiers 
+
+1. All methods on classes / traits are declared `final`, by default.
+2. No methods on objects declared `final`.
+3. No methods on final classes declared `final`.
+4. All classes inside objects should be defined `final`.
+5. All final classes have private constructors & constructor parameters.
+6. All `vals` declared `final`, either in objects or `final classes`, if they are constant expressions and without type annotations.
+7. Private `methods` and `vals` are not declared `final` since they are already final.
+8. Package-private `vals` and `methods` declared `final`.
+
+### Refactoring
+
+1. If a class has all `final` members, the class should be declared `final` and `final` member annotations should be removed except constant expressions.
+2. All type annotations should use the least powerful type alias. This means, that, let us say, a `ZIO` effect that has no dependencies but throws an arbitrary error, should be defined as `IO`.
+
+### Naming of parameters in methods and classes
```

Comment:
Sorry, why do we need this section?

Reply:
Perhaps we do not. Perhaps refactoring is just a dumb name. I admit I will go through everything with a lot more details. Put links to where such rules are based on in the first place. 

Maybe it was too soon to open a PR, this is not ready for a review, it is more for people to take a look, if interested.

Reply:
Ok, fair enough. I'm in general 👎 on exerting this much control on the code people contribute.

Reply:
Most of this I was thinking on enforcing by applying scalafix rules, so, automate it. I just wanted to document it as well. 

I do not want to make things too strict by any means, but create expectations for users reading the code.

Reply:
This would have made sense if it were a commercial team deciding unilaterally that this is how code written in the team should be written. But this is not a commercial team, contributors have different backgrounds and skill levels.

If a reviewer feels a name better suits a method parameter, they can comment on the PR - no need to preemptively specify how the code should be written.

Reply:
What you say makes sense. I will revise my approach on this. My intentions were perhaps too focused on the reader and not the writer.

Reply:
Yes, I would suggest we drop the section on naming. I think we have enough things that are more objective and have a clearer technical rationale that we should focus the time of contributors and reviewers on those.

Reply:
Because I feel I am passing the message I want people to strictly follow these rules and, that was not my intention. I do not want to put any burden on contributors. My intent is the opposite, to make their life easier.

Reply:
I think that is a great attitude. One thing we might want to consider along those lines is to have a brief description of the rationale behind each guideline after it (e.g. "alphabetizing makes it easy for users / contributors to find methods in a large file and avoids wasting time thinking about how to organize things"). That could be a good discipline for us to make sure we have a good justification for these and if people understand it they are more likely to feel this is something helping them versus something they have to do.

Reply:
I do think justifying everything is a good idea. That way we're teaching to think, not just teaching to follow.